File: app/src/main/java/com/valentingonzalez/turistear/activities/login/RegisterActivity.java
package com.valentingonzalez.turistear.activities.login;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.valentingonzalez.turistear.activities.MapsActivity;
import com.valentingonzalez.turistear.R;
import com.valentingonzalez.turistear.includes.BasicToolbar;
import com.valentingonzalez.turistear.models.Usuario;
import com.valentingonzalez.turistear.providers.AuthProvider;
import com.valentingonzalez.turistear.providers.UserProvider;

import dmax.dialog.SpotsDialog;

public class RegisterActivity extends AppCompatActivity {

    AuthProvider mAuthProvider;
    UserProvider mUserProvider;
    MaterialButton mCreateButton;
    TextInputEditText nameTIET;
    TextInputEditText emailTIET;
    TextInputEditText passwordTIET;
    TextInputLayout nameTil;
    TextInputLayout emailTil;
    TextInputLayout passTil;
    AlertDialog progessDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.register_layout);

        mCreateButton = findViewById(R.id.register_button);
        BasicToolbar.show(RegisterActivity.this, "Registrarse", true);

        mAuthProvider = new AuthProvider();
        mUserProvider = new UserProvider();

        nameTIET = findViewById(R.id.nameTiet);
        emailTIET = findViewById(R.id.emailTiet);
        passwordTIET = findViewById(R.id.passTiet);
        nameTil = findViewById(R.id.nameTil);
        emailTil = findViewById(R.id.emailTil);
        passTil = findViewById(R.id.passTil);

        nameTIET.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                nameTil.setError("");
            }
            @Override
            public void afterTextChanged(Editable editable) { }
        });
        emailTIET.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                emailTil.setError("");
            }
            @Override
            public void afterTextChanged(Editable editable) { }
        });
        passwordTIET.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                passTil.setError("");
            }
            @Override
            public void afterTextChanged(Editable editable) { }
        });

        mCreateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                createAccount();
            }
        });
        progessDialog = new SpotsDialog.Builder().setContext(RegisterActivity.this).setMessage("Conectando...").build();
    }

    public void createAccount(){
        final String name = nameTIET.getText().toString();
        final String email = emailTIET.getText().toString();
        String pass = passwordTIET.getText().toString();
        boolean emptyName = false, emptyEmail = false, emptyPass= false, under6Pass= false;
        if(email.isEmpty()){
            emailTil.setError("Email no puede estar vacio");
            emptyEmail = true;
        }
        if(name.isEmpty()){
            nameTil.setError("Nombre no puede estar vacio");
            emptyName = true;
        }
        if(pass.isEmpty()){
            passTil.setError("Contraseña no puede estar vacia");
            emptyPass = true;
        }
        if(pass.length() < 6){
            passTil.setError("Contraseña debe de tener largo mayor a 6");
            under6Pass = true;
        }
        if(emptyName || emptyEmail || emptyPass || under6Pass){
            return;
        }
        progessDialog.show();
        register(email, name, pass);
    }

    public void register(final String email,final String name, String pass){
        mAuthProvider.register(email, pass).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if(task.isSuccessful()){
                    String id = FirebaseAuth.getInstance().getCurrentUser().getUid();
                    Usuario usuario = new Usuario(id, name, email);
                    createUser(usuario);
                }else{
                    Toast.makeText(RegisterActivity.this, "Error con email o password", Toast.LENGTH_SHORT).show();
                }
                progessDialog.dismiss();
            }
        });
    }

    public void createUser(Usuario usuario){
        mUserProvider.createUser(usuario).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful()){
                    Toast.makeText(RegisterActivity.this, "¡Cuenta creada exitosamente!", Toast.LENGTH_SHORT).show();
                    startActivity(new Intent(RegisterActivity.this, MapsActivity.class));
                }else{
                    Toast.makeText(RegisterActivity.this, "Error con creacion de usuario", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }
}

File: app/src/main/java/com/valentingonzalez/turistear/activities/login/RegisterActivity.kt
package com.valentingonzalez.turistear.activities.login

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.valentingonzalez.turistear.R
import com.valentingonzalez.turistear.activities.MapsActivity
import com.valentingonzalez.turistear.includes.BasicToolbar.show
import com.valentingonzalez.turistear.models.Usuario
import com.valentingonzalez.turistear.providers.AuthProvider
import com.valentingonzalez.turistear.providers.UserProvider
import dmax.dialog.SpotsDialog

class RegisterActivity : AppCompatActivity() , UserProvider.UserProviderListener{
    var mAuthProvider: AuthProvider? = AuthProvider()
    var mUserProvider: UserProvider = UserProvider(this)
    var mCreateButton: MaterialButton? = null
    var nameTIET: TextInputEditText? = null
    var emailTIET: TextInputEditText? = null
    var passwordTIET: TextInputEditText? = null
    var nameTil: TextInputLayout? = null
    var emailTil: TextInputLayout? = null
    var passTil: TextInputLayout? = null
    var progessDialog: AlertDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.register_layout)
        mCreateButton = findViewById(R.id.register_button)
        show(this@RegisterActivity, "Registrarse", true)
        nameTIET = findViewById(R.id.nameTiet)
        emailTIET = findViewById(R.id.emailTiet)
        passwordTIET = findViewById(R.id.passTiet)
        nameTil = findViewById(R.id.nameTil)
        emailTil = findViewById(R.id.emailTil)
        passTil = findViewById(R.id.passTil)
        nameTIET?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                nameTil?.error = ""
            }
            override fun afterTextChanged(editable: Editable) {}
        })
        emailTIET?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                emailTil?.error = ""
            }
            override fun afterTextChanged(editable: Editable) {}
        })
        passwordTIET?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                passTil?.error = ""
            }
            override fun afterTextChanged(editable: Editable) {}
        })
        mCreateButton?.setOnClickListener{ createAccount() }
        progessDialog = SpotsDialog.Builder().setContext(this@RegisterActivity).setMessage("Conectando...").build()
    }
    fun createAccount() {
        val name = nameTIET!!.text.toString()
        val email = emailTIET!!.text.toString()
        val pass = passwordTIET!!.text.toString()
        var emptyName = false
        var emptyEmail = false
        var emptyPass = false
        var under6Pass = false
        if (email.isEmpty()) {
            emailTil!!.error = "Email no puede estar vacio"
            emptyEmail = true
        }
        if (name.isEmpty()) {
            nameTil!!.error = "Nombre no puede estar vacio"
            emptyName = true
        }
        if (pass.isEmpty()) {
            passTil!!.error = "Contraseña no puede estar vacia"
            emptyPass = true
        }
        if (pass.length < 6) {
            passTil!!.error = "Contraseña debe de tener largo mayor a 6"
            under6Pass = true
        }
        if (emptyName || emptyEmail || emptyPass || under6Pass) {
            return
        }
        progessDialog!!.show()
        register(email, name, pass)
    }

    fun register(email: String?, name: String?, pass: String?) {
        mAuthProvider!!.register(email, pass).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val id = FirebaseAuth.getInstance().currentUser!!.uid
                val usuario = Usuario(id, name, email)
                createUser(usuario)
            } else {
                Toast.makeText(this@RegisterActivity, "Error con email o password", Toast.LENGTH_SHORT).show()
            }
            progessDialog!!.dismiss()
        }
    }

    fun createUser(usuario: Usuario?) {
        mUserProvider.createUser(usuario!!).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                Toast.makeText(this@RegisterActivity, "¡Cuenta creada exitosamente!", Toast.LENGTH_SHORT).show()
                startActivity(Intent(this@RegisterActivity, MapsActivity::class.java))
            } else {
                Toast.makeText(this@RegisterActivity, "Error con creacion de usuario", Toast.LENGTH_SHORT).show()
            }
        }
    }
    override fun onFavoriteChecked(isFav: List<Boolean>) {
    }
    override fun getUserName(user: Usuario) {
    }
}